design/authenticators/authenticators_as_external_plugins.md

```diff
@@ -0,0 +1,36 @@
+# Solution Design - Authenticator as a Plugin
+
+## Table of Contents
+
+- [Problem Statement](#problemstatement)
+- [Goals](#goals)
+- [Implementation Steps](#implementationsteps)
+
+
+## Problem Statement
+
+Currently adding new authenticator type in Conjur / DAP requires code changes in Conjur.
+
+Thus is only possible to external users by code contribution to Conjur, and impossible for a concrete DAP version (which uses predefined Conjur version)
+
+Adding an ability to implement authenticator as an external plugin will allow our customers to implement authenticators of their own without changing Conjur codebase 
+
+and without being familiar with Ruby / Conjur
+
+## Goals
+Create programmatic / configurational hooks in Conjur, which will allow to incorporate external authenticator as a plugin to Conjur
+Make authenticator addition to be possible for customers
+Preserve set of authenticators that exist in Conjur
+Provide set of parameters for plugin-based authenticators per user decision
+
+## Implementation Steps
+- Identify the functional blocks that are common to all authenticators and split them into 2 kinds
+  Blocks that are not going to be programmed by plugin (example: set / get authenticator in DB)
+  Blocks that should be programmed by plugin (example: validate methods)
+- Verify that base blocks inside Conjur are well separated from authenticator-specific blocks inside Conjur. 
+- If they are not - take steps to make them such
+- Define hook to load external authenticators
+- Define hook to provide external authenticator parameters (probably in yaml)
+- Define hook for "validate" step of external authenticator (should contain a pointer to function / executable)
+- Verify that configuration of external authenticator is possible in Conjur
+- Implement and test sample external authenticator
```

Comment:
This approach is unlikely to be flexible enough to handle all possible future authenticators.  Authentication schemes differ enough that we don't want to go for a "function template" solution:  That is, we can't find the steps that "all" authenticators take, make them into into abstract methods, and have each authenticator "fill them in".

Something like the original design that Kevin proposed seems more robust to me:

In his proposal, an authenticator was just any dockerized webservice, which could be configured to be "attached" to Conjur.  That webservice could have whatever API it wanted, including multiple routes if needed, and could be written in any language.  The webservice could then "mint" new conjur tokens at will, by requesting them from the core conjur service via a unix websocket.  

Thus the basic architecture is:

1. A generic webservice with whatever API it needed which is responsible for determining when something is authenticated.
2. A mechanism (provided by Conjur core) for this webservice to create a conjur token once it decides something is authenticated.

This makes no assumption about the generic "structure" of an authenticator, and this is good, because it is unlikely any such abstraction will stand up to the test of time.

Reply:
Thanks @jonahx - so essentially we will have an authentication service decoupled from the core project
Going that direction, I wonder what would you think on generating the tokens by this service and have conjur just validate them.
@oburstein-hub @orenbm - WDYT?

Reply:
Generating tokens sounds a good idea, but it should be detailed in the design

Reply:
@Tovli In Kevin's design, conjur-core was still responsible for generating the tokens (so that only conjur-core had access to the signing key), but the webservice plugins could ask conjur to generate tokens for them, via a local unix socket.  The local unix socket isn't the only way to do this, ofc, but I like the idea that minting tokens remains the centralized responsibility of conjur-core.

Reply:
@jonahx before we drop the suggested approach, i want to clarify that the authentication flow can be separated into 2 parts:
  - General validations
  - Specific validations

This can be seen in the `call` function of [`Authentication:: Authenticate`](https://github.com/cyberark/conjur/blob/4a02078f7921d195f1e02cf7070041d4c1b05ebc/app/domain/authentication/authenticate.rb):
```
def call
  validate_authenticator_exists
  validate_webservice_is_whitelisted
  validate_user_has_access_to_webservice
  validate_origin
  validate_credentials
  audit_success
  new_token
rescue => e
  audit_failure(e)
  raise e
end
```

The authentication performs all the check until `validate_credentials`, which is implemented as follows:
```
def validate_credentials
  raise Errors::Authentication::InvalidCredentials unless authenticator.valid?(@authenticator_input)
end
```

The `valid?` function is implemented inside the authenticator class and it performs the specific authentication (e.g verify the ID token in OIDC).

So, the way i see it, a new authenticator needs only to implement the specific validations of the new authenticator. I created such a template just to pass the idea (it is not implemented as a plugin). Please see the classes i added [here](https://github.com/cyberark/conjur/tree/authenticator-template/app/domain/authentication/template).

We can change the way that we build the `authenticators` array that has all the authenticator classes so that these classes can be loaded in runtime instead of compile time.

Won't this approach be generic enough for us?

Reply:
i am thinking of something similar to what we have in secretless but i'm not sure we have this in Ruby

Reply:
As I said over chat, it all depends on what the goal is.  From my understanding, being able to write these in any language would be a huge boon -- to the sales engineers, to customers, and for the community efforts.  So if we take that as a given, then our current approach can't support that.

I don't think there's anything wrong with our current approach.  It's simple and I helped design it.  But I think the way things have played out, it's become clear that "plugins you can write in any language" is probably where we ultimately want to go.